### ISSUE TITLE
[mu bug] Failed to open glass revision file for reading

### ISSUE THREAD

**Issue by haji-ali** [djcb/mu#2144]
**Describe the bug**
Occasionally (once a week or so), I get the error
```
error in process filter: [mu4e] Error 111: failed to open store @ ~/.cache/mu/xapian: ~/.cache/mu/xapian/iamglass: Failed to open glass revision file for reading
```
when using `mu4e` or running `mu` with any command.

Looking in `~/.cache/mu/xapian/` I only find one file: `flintlock`. I can fix the issue by running `mu init` and `mu index` (and waiting 15 minutes or so to index my 35k messages). But having to do this every week is not tenable.

I don't manually manipulate the xapian files and I only use xapian through mu, so I am not sure what is deleting the `iamglass` file.

**To Reproduce**
I am unable to find the exact conditions to reproduce this error systematically, except to say that it happens once a week or so.

This didn't seem to happen in old version of mu (like 1.5) and it seems to have started in 1.6, but I am not 100% sure.
I am happy to follow any steps to troubleshoot this issue and provide more info.

**Environment**
mu: 1.7
OS: CentOS 7

**Checklist**
- [x] you are running either the latest 1.6.x release, or a 1.7.x development release (otherwise, please upgrade).

**Comment by theophilusx** [comment]
Abdul-Lateef Haji-Ali ***@***.***> writes:

> Describe the bug
> Occasionally (once a week or so), I get the error
>
> error in process filter: [mu4e] Error 111: failed to open store @ ~/.cache/mu/xapian: ~/.cache/mu/xapian/iamglass: Failed to open glass revision file for reading
>
> when using mu4e or running mu with any command.
>
> Looking in ~/.cache/mu/xapian/ I only find one file: flintlock. I can fix the issue by running mu init and mu index (and
> waiting 15 minutes or so to index my 35k messages). But having to do this every week is not tenable.
>
> I don't manually manipulate the xapian files and I only use xapian through mu, so I am not sure what is deleting the iamglass
> file.
>
> To Reproduce
> I am unable to find the exact conditions to reproduce this error systematically, except to say that it happens once a week or so.
>
> This didn't seem to happen in old version of mu (like 1.5) and it seems to have started in 1.6, but I am not 100% sure.
> I am happy to follow any steps to troubleshoot this issue and provide more info.
>
> Environment
> mu: 1.7
> OS: CentOS 7
>
> Checklist
>
> *  you are running either the latest 1.6.x release, or a 1.7.x development release (otherwise, please upgrade).
>

I think the flintlock file is a simple lock file used to ensure only one
process is accessing/updating the xapian store at a time. I've seen this
error after I've exited emacs abruptly when it was in the middle of an
index update. To fix it, I just delete the file (after ensuring no
process is 'stuck' trying to update the store).

I guess that if you have configured mu4e to automatically update in the
background every x minutes (which is what I do), it is possible that you
may quit Emacs while it is in the middle of an update, which would
prevent the lock file being removed as it should.

**Comment by haji-ali** [comment]
@theophilusx, yes that's possible (although I rarely quit Emacs). The error I typically see when the lockfile is not deleted is slightly different
```
error: failed to open store @ /home/ah180/.cache/mu/xapian: Unable to get write lock on /home/ah180/.cache/mu/xapian: already locked
```
The error I mentioned above is different and is related to the missing `*glass` files.

I could try deleting the lock file the next time this happens, but the problem is that the `*glass` files are deleted; I assume these are the index files and would need to be regenerated (taking 15-20 minutes I presume).

**Comment by djcb** [comment]
_Something_ seems to be deleting the cache file on your system, perhaps a cron job or something like that? Otherwise, you can try to move mu's home directory elsewhere (outside ~/.cache) and see if that helps.

(btw 15 minutes for 35K seems rather slow)

**Comment by haji-ali** [comment]
> _Something_ seems to be deleting the cache file on your system, perhaps a cron job or something like that? Otherwise, you can try to move mu's home directory elsewhere (outside ~/.cache) and see if that helps.

I am not sure what would be deleting the cache file besides mu. But I tried moving `muhome` to another directory and will check if that helps.

BTW, I found it strange that I need to pass the home directory to `mu index` since I already set that up for `mu init`, is there a reason this setting is not saved as other settings are?


> (btw 15 minutes for 35K seems rather slow)

I seem to have overestimated this time. After bench-marking, it takes 8-9 minutes. 
Not sure if this is within reason I experience regular slow downs of `mu index` (I previously reported on this [here](https://github.com/djcb/mu/issues/2098#issuecomment-902789225))

**Comment by djcb** [comment]
You need to tell mu where its home is (when it's not default), otherwise it cannot find those (mu-home-specific) settings.

For comparison, I'm indexing ~95K messages (after `mu init`) in under 4 minutes. So perhaps 8 minutes is within range for a slightly slower machine.

Anyway, just can use the "old" ~/.cache mu, index and leiave it alone a bit and see if it's deleted. Perhaps your system log has some info.

**Comment by benjamin-asdf** [comment]
You also get this error before the first `mu init`
